### PR TITLE
Ignore padding tokens when generating text

### PR DESCRIPTION
**Pull request by gcooper-isi**
To ensure that the vocabulary can be evenly divided across all the GPUs that the model is split across, and also apparently for some efficiency reasons I don't fully understand, MegatronLM adds padding tokens to the model. But these padding tokens do not correspond to any real tokens in the vocabulary, and thus can't even be decoded. The current version of the text generation code allows these tokens to be selected during generation, which causes an exception to be thrown.

This corresponds to one of the two issues originally fixed in https://github.com/gcooper-isi/DeepSpeedExamples/pull/19/files

NVIDIA has already fixed the other issue addressed in that other pull request (that generation was only using the tokens for the first GPU within each data parallel group when generating text).

### CODE REVIEW

**Comment by wrichardson-isi** [comment]
I've successfully trained a model with this branch and then used that model to generate text successfully.  I'm generating a few more samples right now.  Do you have any further guidance as to what might be good testing steps here?

**Comment by gcooper-isi** [comment]
This one is not so easy to test, since there's no way to guarantee that the exception that we're worried about will occur. But it tends to happen when you train a toy model with an extremely small number of parameters, which I sometimes do just to be able to quickly test out the mechanics of things. So you can do that to try and trigger the bug. If you can, try running generation twice with the same seed, once on master and once on the dont_generate_with_padding_tokens branch. If you can get the error to occur on mater but not the dont_generate_with_padding_tokens branch that's a good indication that it worked.

The other thing to do is to print out the size of the logits tensor before and after it gets truncated. Before, the last dimension should have a size equal to the total number of tokens including the padding tokens, and after, the last dimension should have a size equal to just the number of "real" tokens. The generation job should print out both of those numbers for reference.

**Comment by wrichardson-isi** [comment]
I'm ready to sign off on this.  It's working correctly.
